packages/protocol/contracts/dapis/rrpDapi.sol

```diff
@@ -2,13 +2,91 @@
 pragma solidity 0.8.4;
 
 import "./reducers/Reducer.sol";
+import "./rrpDapiServer.sol";
 
 // A stub for an RRP dAPI contract
 // When the user wants to deploy a dAPI contract, they inherit the specific
 // reducer contract such as:
 //       contract myRrpDapi is rrpDapi, Median {
 //           // Any specific implementation required
 //       }
-abstract contract rrpDapi is Reducer {
-    // TODO: Implement
+abstract contract RrpDapi is Reducer {
+
+    RrpDapiServer public rrpDapiServer;
+
+    uint256 public immutable maxAnswersInStorage;
+    uint256 private immutable dapiId;
+    uint256 private immutable requesterIndex;
+
+    uint256 public nextAnswerId = 1;
+    
+    mapping(uint256 => uint256) private requestIndexToAnswerIdMod;    
+    mapping(uint256 => uint256) private answerIdModToRequestIndex;    
+    mapping(uint256 => int256) private answerIdModToValue;    
+
+    constructor
+    (
+      address _rrpDapiServer,
+      uint256 _requesterIndex,
+      uint256 _maxAnswersInStorage,
+      uint256 _minResponsesToReduce,
+      int256 _toleranceInPercentages,
+      bytes32[] memory _templateIds,
+      address[] memory _designatedWallets
+    ){
+      rrpDapiServer = RrpDapiServer(_rrpDapiServer);
+      
+      maxAnswersInStorage = _maxAnswersInStorage;
+      requesterIndex = _requesterIndex;
+      
+      // register with the main RRP dAPI server
+      dapiId = rrpDapiServer.registerDapi(
+        _minResponsesToReduce,
+        _toleranceInPercentages,
+        _requesterIndex,
+        _templateIds,
+        _designatedWallets,
+        address(this),
+        this.reduce.selector
+      );
+    }
+
+    function getAnswer(uint answerId) internal view returns (int256) {
+      require(
+             0 < answerId
+             && (answerId <= maxAnswersInStorage || nextAnswerId - maxAnswersInStorage <= answerId)
+             && answerId < nextAnswerId,
+            "Answer doesn't exist."
+        );
+        return answerIdModToValue[answerId % maxAnswersInStorage + 1];
+    }
+
+    function makeRequest() external returns (uint256 answerId){
+      
+      // reverts if requester is not endorsed
+      uint256 requestIndex = rrpDapiServer.makeDapiRequest(dapiId, requesterIndex);
+
+      // +1 is added to the modulo to prevent 0 values
+      uint256 answerMod = nextAnswerId % maxAnswersInStorage + 1;
+      
+      // delete stale answer in storage
+      delete requestIndexToAnswerIdMod[answerIdModToRequestIndex[answerMod]];
+      
+      requestIndexToAnswerIdMod[requestIndex] = answerMod;
```

Comment:
This writes to an unwritten storage location for each request. This is in general is not more efficient than keeping the answers in an infinitely growing array.

Reply:
Even if you delete the previous one? Are you 100% sure?

Reply:
If so, I can still implement a constant storage, it's similar to this algorithm: https://www.geeksforgeeks.org/how-to-delete-an-element-from-the-set-by-passing-its-value-in-c/

Reply:
`requestIndex` is constantly being incremented so yeah. The main point is that overwriting the reduced answer is not worth it due to these overheads, this needs to be done for individual responses.

Reply:
Also this should be implemented with the assumption that `delete` won't refund gas due to EIP-3300

Reply:
I'm not sure I understand your point about overhead? Writing to allocated, fixed-sized storage will be more gas efficient then an ever-increasing storage array.

Reply:
> Writing to allocated, fixed-sized storage will be more gas efficient then an ever-increasing storage array.

Only if you don't have to write additional stuff (timestamp to check expiry for example) to storage to make sure that you're not overwriting answers you shouldn't overwrite. If it's a data feed and you're overwriting an `int256` then it will be very gas efficient.